### PR TITLE
OpenELM support

### LINKED ISSUE #6868: Support for OpenELM of Apple
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [ ] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [ ] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [ ] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [ ] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Feature Description

Support for OpenELM of Apple

https://huggingface.co/apple/OpenELM-3B-Instruct/tree/main

### PR DESCRIPTION
**Pull request by icecream95**
Fixes: #6868.

Thanks to @joshcarp for an initial try at doing this (#6986), it was very helpful as a source to copy-paste from and check against.

Currently a bunch of the configuration is hardcoded into `llama.cpp`, so only the 270M model works at this point.

The `ffn_up` tensors in the converted model are actually concatenations of `ffn_gate` and `ffn_up`, perhaps the conversion script should separate them out?

The 270M model is impressively fast, and works fine for generation, but "Chat" mode in `./server` doesn't really work well. Perhaps that's just because it hasn't been finetuned for that? I'm not really sure.

### CODE REVIEW

**Comment by icecream95** [comment]
It looks like context shift currently causes crashes, because `build_k_shift` uses the false number of heads in the `.gguf`.

A few other functions seem like they will be broken as well.

**Comment by icecream95** [comment]
I've been quite tired recently, so it might be a while before I'm able to come back to this.

I see that @jart's #7445 has already been merged with similar modifications to `llama_model_type_name`, but I think git merge will do the right thing here without needing to change that commit.

**Comment by joshcarp** [comment]
@icecream95 I might jump back on this, curious to where i got stuck

**Comment by joshcarp** [comment]
@icecream95 might jump back on this cause im curious to where i got stuck

**Comment by compilade** [comment]
> We'll probably need to generalize the head number to be determined per layer.

I'll be working on porting the variable GQA stuff written for Jamba (in #7531) to OpenELM in the next days, which will both help with making OpenELM work and with reducing the size of the Jamba PR (ref: <https://github.com/ggerganov/llama.cpp/pull/7531#issuecomment-2173483542>).

I think the `ffn_multipliers` could also be simplified similarly by taking care of them in the convert script while the resulting FFN sizes could be directly part of `{arch}.feed_forward_length` as an array of integers.

@icecream95 do you mind if I push to this branch, or would you prefer that I use another one? In the meantime I'll work on a local branch.

**Comment by compilade** [comment]
I've got this working for the 270M model, the 1.1B model, and the 450M model. Getting reasonable perplexity. I did not yet test the 3B model, but will try once it finishes downloading.

According to <https://github.com/apple/corenet/blob/2261885b6696950aaf481a862e8926921ef1a067/projects/openelm/instruction_tuning/openelm-instruct.yaml#L7>, the prompt template is zephyr, but it seems to perform poorly in practice, while an alpaca-like template (with `### Instruction:`) seems to work better for some reason.

This might be because the tokenizer doesn't have tokens for `<|user|>` and `<|assistant|>`, since it's using Llama-2's tokenizer.

File: src/llama.cpp [hunk 1]
```diff
@@ -6777,6 +6861,41 @@ static bool llm_load_tensors(
...
+                    for (int i = 0; i < n_layer; ++i) {
+                        const int64_t n_head = hparams.n_head_l(i);
+                        const int64_t n_head_qkv = 2*hparams.n_head_kv_l(i) + n_head;
+                        const int64_t n_embd_head = hparams.n_embd_head_k;
+
+                        const int64_t n_ff = hparams.n_ff_l(i);
+
+                        ggml_context* ctx_layer = ctx_for_layer(i);
+                        ggml_context* ctx_split = ctx_for_layer_split(i);
+                        auto & layer = model.layers[i];
```
Comment (ggerganov) [inline src/llama.cpp:6884]:
```suggestion
                        ggml_context * ctx_layer = ctx_for_layer(i);
                        ggml_context * ctx_split = ctx_for_layer_split(i);

                        auto & layer = model.layers[i];
```

File: src/llama.cpp [hunk 2]
```diff
@@ -2092,6 +2112,11 @@ struct llama_hparams {
     uint32_t n_expert_used = 0;
     uint32_t n_vocab_type = 0; // for BERT-style token types
 
+    // TODO: find a more compact way to add more per-layer hyper-parameters
+    std::vector<int32_t> n_head_vec;
+    std::vector<int32_t> n_head_kv_vec;
+    std::vector<int32_t> n_ff_vec;
```
Comment (ggerganov) [inline src/llama.cpp:2118]: Do we expect negative values here? If not, we should change to `uint32_t`

File: src/llama.cpp [hunk 3]
```diff
@@ -2173,18 +2202,53 @@ struct llama_hparams {
         return false;
     }
 
-    uint32_t n_gqa() const {
+    // TODO: deduplicate per-layer getters
+    uint32_t n_head_l(uint32_t layer) const {
```
Comment (ggerganov) [inline src/llama.cpp:2206]: We can remove `hparams.n_head` and rename this to:

```c++
uint32_t n_head(uint32_t layer) const {
```

This way everywhere we reference `hparams.n_head`, we will use `hparams.n_head()`.
We can do this for the rest of the parameters that make sense to be per-layer, the main goal being to avoid having duplicated information in `hparams.n_head` and `hparams.n_head_vec`.

We can do this change in a follow-up refactoring PR, together with finding a more compact way to have per-layer hparams (maybe via new `struct llama_hparams_layer`)

**Review by ggerganov — APPROVED** [review]
> This might be because the tokenizer doesn't have tokens for <|user|> and <|assistant|>, since it's using Llama-2's tokenizer.

Do we know what is the reason for the tokenizer confusion? It seems the model is unusable without the correct tokenizer

File: src/llama.cpp (hunk 2 above)
Comment (compilade) [reply src/llama.cpp:2118]: I've used `int32_t` here due to a limitation of `GGUFWriter`, which doesn't allow choosing the specific type of integer arrays and defaults to `INT32`. And also because `get_arr` in `src/llama.cpp` can only load arrays into vectors which have exactly the same type as the loaded array.

I'll see if this can be fixed, but it will likely require using Numpy types (e.g. `np.uint32`) in `GGUFValueType.get_type()` in `gguf-py/gguf/constants.py`.

**Comment by compilade** [comment]
> Do we know what is the reason for the tokenizer confusion? It seems the model is unusable without the correct tokenizer

I'm not sure. They clearly say [they used the Llama 2 tokenizer](https://github.com/apple/corenet/tree/main/projects/openelm#tokenizer), while they also [suggest to use it verbatim](https://github.com/apple/corenet/blob/2261885b6696950aaf481a862e8926921ef1a067/mlx_examples/open_elm/README.md?plain=1#L27-L29). This doesn't feel right, but this is apparently what they have done.

From playing around with the OpenELM Instruct models with only the BOS token as a prompt and different seeds at `--temp 1`, they show no sign of having been trained on a chat template.

When asking questions or instructions, at least they *seem* to follow them, but the answers are never short (which can be useful in some cases).

It's a bit unfortunate they did not add special tokens (or even re-use the BOS and EOS token in a chatml-style way, like the bagel models do), which makes their instruct models worse than they could have been. Although this could likely be fixed by better fine-tuning.

File: src/llama.cpp (hunk 3 above)
Comment (compilade) [reply src/llama.cpp:2206]: The advantage of duplicating the information is to have a fallback value, and it also avoids having to allocate the vectors when there's only one value for all the layers. The disadvantage is the possible confusion of which value to use.

> We can do this change in a follow-up refactoring PR, together with finding a more compact way to have per-layer hparams (maybe via new `struct llama_hparams_layer`)

Yes a follow-up refactor seems appropriate. The goal with this initial implementation was to minimize having to change the accesses to `hparams.n_head`, `hparams.n_head_kv`, and `hparams.n_ff` all over the place, and also to re-use the nice abstraction which `llama_model_loader.get_arr` seemed to be (it was introduced in #7225, but isn't used by anything on `master`).

Although a possible huge problem with using `std::vector` in `llama_hparams` is that it's no longer trivially copyable as assumed by `llama_state_save_file_internal` and `llama_state_load_file_internal`.

Comment (ggerganov) [reply src/llama.cpp:2206]: > Although a possible huge problem with using std::vector in llama_hparams is that it's no longer trivially copyable as assumed by llama_state_save_file_internal and llama_state_load_file_internal.

Hm, correct. We could switch to `uint32 n_head_vec[LLAMA_MAX_LAYERS];` instead and use `n_layer` as the container size?

**Comment by sqzhang-jeremy** [comment]
> I've got this working for the 270M model, the 1.1B model, and the 450M model. Getting reasonable perplexity. I did not yet test the 3B model, but will try once it finishes downloading. EDIT: the 3B model works too!
> 
> According to https://github.com/apple/corenet/blob/2261885b6696950aaf481a862e8926921ef1a067/projects/openelm/instruction_tuning/openelm-instruct.yaml#L7, the prompt template is zephyr, but it seems to perform extremely poorly in practice, while an alpaca-like template (with `### Instruction:`) seems to work better for some reason. Even no template at all works pretty well from what I've seen.
> 
> This might be because the tokenizer doesn't have tokens for `<|user|>` and `<|assistant|>`, since it's using Llama-2's tokenizer.
> 
> Session file saving support isn't yet done, but everything else should pretty much work.

Question: How did you get the .gguf format OpenELM models? Could you please share the experience? Thanks!

File: src/llama.cpp [hunk 4]
```diff
@@ -4371,9 +4500,20 @@ static void llm_load_hparams(
...
 
+    // zero-out the per-layer hparams
+    std::fill(hparams.n_head_arr.begin(),    hparams.n_head_arr.end(),    0);
+    std::fill(hparams.n_head_kv_arr.begin(), hparams.n_head_kv_arr.end(), 0);
+    std::fill(hparams.n_ff_arr.begin(),      hparams.n_ff_arr.end(),      0);
+
+    ml.get_key_or_arr(LLM_KV_FEED_FORWARD_LENGTH,  hparams.n_ff_arr,   hparams.n_layer);
+    ml.get_key_or_arr(LLM_KV_ATTENTION_HEAD_COUNT, hparams.n_head_arr, hparams.n_layer);
+
+    GGML_ASSERT(hparams.n_head() > 0);
```
Comment (compilade) [inline src/llama.cpp:4535]: @ggerganov (from your change in <https://github.com/ggerganov/llama.cpp/commit/29ab5a0ed121cdbcad3012e07154a6d3132e01a6>) This will fail for non-Transformer models like Mamba, and even partially non-Transformer models like Jamba.

(This was also the reason for the checks before divisions)

Comment (ggerganov) [reply src/llama.cpp:4535]: Thank, I forgot about this. Will try to resolve

**Comment by ggerganov** [comment]
@compilade Shall we merge after green CI?

**Review by compilade — APPROVED** [review]
> Shall we merge after green CI?

Yes, this feels pretty much ready.

**Comment by ngxson** [comment]
I've been trying some chat templates, but there is no sign that it is trained for chat.

Apparently, apple did not make it clear if "Instruction" means single turn Q&A or multi-turn conversation. Hopefully someone will release a fine tune with proper chat template support.

**Comment by sqzhang-jeremy** [comment]
Hi there!

I tried to deploy the openELM series models like openELM. After transforming to .gguf format , I met a bug which showed that unknown architecture openelm

Platform: Android Pixel 8 Pro(ARM)
Git Hash: a38b884c6c4b0c256583acfaaabdf556c62fabea

Q: How to tackle load model issue?
`~/llama.cpp/bin $ ./llama-cli -m ../../model/apple/openelm-270M-it-model-q8_0.gguf -n 128 --color -if
Log start
main: build = 3291 (f6190247)
main: built with clang version 18.1.8 for aarch64-unknown-linux-android24
main: seed  = 1720242044
llama_model_loader: loaded meta data with 25 key-value pairs and 146 tensors from ../../model/apple/openelm-270M-it-model-q8_0.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = openelm
llama_model_loader: - kv   1:                               general.name str              = openelm-270M
llama_model_loader: - kv   2:                        openelm.block_count u32              = 16
llama_model_loader: - kv   3:                     openelm.context_length u32              = 2048
llama_model_loader: - kv   4:                   openelm.embedding_length u32              = 1280
llama_model_loader: - kv   5:                openelm.feed_forward_length arr[i32,16]      = [768, 1024, 1280, 1536, 1792, 2048, 2...
llama_model_loader: - kv   6:               openelm.attention.head_count arr[i32,16]      = [12, 12, 12, 12, 12, 16, 16, 16, 16, ...
llama_model_loader: - kv   7:            openelm.attention.head_count_kv arr[i32,16]      = [3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, ...
llama_model_loader: - kv   8:                     openelm.rope.freq_base f32              = 10000.000000
llama_model_loader: - kv   9:   openelm.attention.layer_norm_rms_epsilon f32              = 0.000001
llama_model_loader: - kv  10:               openelm.rope.dimension_count u32              = 64
llama_model_loader: - kv  11:               openelm.attention.key_length u32              = 64
llama_model_loader: - kv  12:             openelm.attention.value_length u32              = 64
llama_model_loader: - kv  13:                          general.file_type u32              = 7
llama_model_loader: - kv  14:                       tokenizer.ggml.model str              = llama
llama_model_loader: - kv  15:                         tokenizer.ggml.pre str              = default
llama_model_loader: - kv  16:                      tokenizer.ggml.tokens arr[str,32000]   = ["<unk>", "<s>", "</s>", "<0x00>", "<...
llama_model_loader: - kv  17:                      tokenizer.ggml.scores arr[f32,32000]   = [0.000000, 0.000000, 0.000000, 0.0000...
llama_model_loader: - kv  18:                  tokenizer.ggml.token_type arr[i32,32000]   = [2, 3, 3, 6, 6, 6, 6, 6, 6, 6, 6, 6, ...
llama_model_loader: - kv  19:                tokenizer.ggml.bos_token_id u32              = 1
llama_model_loader: - kv  20:                tokenizer.ggml.eos_token_id u32              = 2
llama_model_loader: - kv  21:            tokenizer.ggml.unknown_token_id u32              = 0
llama_model_loader: - kv  22:               tokenizer.ggml.add_bos_token bool             = true
llama_model_loader: - kv  23:               tokenizer.ggml.add_eos_token bool             = false
llama_model_loader: - kv  24:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:   65 tensors
llama_model_loader: - type q8_0:   81 tensors
llama_model_load: error loading model: error loading model architecture: unknown model architecture: 'openelm'
llama_load_model_from_file: failed to load model
llama_init_from_gpt_params: error: failed to load model '../../model/apple/openelm-270M-it-model-q8_0.gguf'
main: error: unable to load model`

**Comment by compilade** [comment]
@sqzhang-jeremy

> How to tackle load model issue?

Try to rebuild your binaries.

> `Log start main: build = 3291 (f6190247)`

This build is from a commit from before OpenELM support was merged.

**Comment by sqzhang-jeremy** [comment]
> @sqzhang-jeremy Thanks for trying this.
> 
> > How to tackle load model issue?
> 
> Try to rebuild your binaries.
> 
> > `Log start main: build = 3291 (f6190247)`
> 
> This build is from a commit from before OpenELM support was merged.

After rebuilding, It worked! Thank you!
